api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListChildrenOnARTandFormulationsDataset.java

```diff
@@ -1,24 +1,784 @@
+            + "                 INNER JOIN obs o  "
+            + "                     ON e.encounter_id = o.encounter_id  "
+            + "             WHERE  p.voided = 0  "
+            + "                 AND e.voided = 0  "
+            + "                 AND o.voided = 0  "
+            + "                 AND e.location_id = :location "
+            + "                 AND e.encounter_type IN (${6}, ${9}) "
+            + "                 AND e.encounter_datetime <= :endDate "
+            + "             GROUP BY p.patient_id "
+            + "                   ) most_recent  ON p.patient_id = most_recent.patient_id   "
+            + " WHERE  p.voided = 0  "
+            + "     AND e.voided = 0  "
+            + "     AND o.voided = 0  "
+            + "     AND e.location_id = :location "
+            + "     AND e.encounter_type IN (${6}, ${9}) "
+            + "     AND e.encounter_datetime <= :endDate "
+            + "     AND e.encounter_datetime = most_recent.e_datetime "
+            + "     AND o.concept_id = ${23739} "
+            + "     AND o.value_coded = ${23720} ";
+
+    StringSubstitutor substitutor = new StringSubstitutor(valuesMap);
+
+    spdd.setQuery(substitutor.replace(sql));
+    return spdd;
+  }
+
+  /**
+   * 18
+   *
+   * @return
+   */
+  private DataDefinition getNextFollowUpConsultationDate() {
+    SqlPatientDataDefinition spdd = new SqlPatientDataDefinition();
+    spdd.setName("Next Follow up Consultation Date");
+    spdd.addParameter(new Parameter("location", "location", Location.class));
+    spdd.addParameter(new Parameter("endDate", "endDate", Date.class));
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("6", hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("9", hivMetadata.getPediatriaSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("1410", hivMetadata.getReturnVisitDateConcept().getConceptId());
+
+    String sql =
+        " SELECT p.patient_id, MAX(o.value_datetime) "
+            + " FROM   patient p  "
+            + "         INNER JOIN encounter e  "
+            + "                         ON p.patient_id = e.patient_id  "
+            + "         INNER JOIN obs o  "
+            + "                         ON e.encounter_id = o.encounter_id  "
+            + " WHERE  p.voided = 0  "
+            + "         AND e.voided = 0  "
+            + "         AND o.voided = 0  "
+            + "         AND e.location_id = :location "
+            + "         AND e.encounter_type IN (${6}, ${9}) "
+            + "         AND o.concept_id = ${1410}  "
+            + "         AND e.encounter_datetime <= :endDate "
+            + " GROUP BY p.patient_id ";
+
+    StringSubstitutor substitutor = new StringSubstitutor(valuesMap);
+
+    spdd.setQuery(substitutor.replace(sql));
+    return spdd;
+  }
+
+  /**
+   * 9
+   *
+   * @return
+   */
+  private DataDefinition getFormulation1() {
```

Comment:
maybe we can simplify all formulations into one method and separate calls using a flag as parameter, same recommendation for the calculation itself, just use one class instead of 4 separate.

Reply:
done